File: backend/src/services/announcementService.js
import { getHeadcountReport } from "./headcountService.js";

const MEAL_LABELS = {
    LUNCH: "Lunch",
    SNACKS: "Snacks",
    IFTAR: "Iftar",
    EVENT_DINNER: "Event Dinner",
    OPTIONAL_DINNER: "Optional Dinner",
};

function formatDate(dateStr) {
    // Parse manually to avoid timezone shifts
    const [y, m, d] = dateStr.split("-").map(Number);
    return new Date(y, m - 1, d).toLocaleDateString("en-GB", {
        weekday: "long",
        day: "numeric",
        month: "short",
        year: "numeric",
    });
}

function specialDayLabel(type) {
    return type
        .split("_")
        .map((w) => w.charAt(0) + w.slice(1).toLowerCase())
        .join(" ");
}

/**
 * Generates a markdown announcement string for the given date.
 * Matches the sample format in technical-design.md (lines 596-624).
 */
export function generateAnnouncement(date) {
    const report = getHeadcountReport(date);
    const formattedDate = formatDate(date);
    const lines = [];

    lines.push(`**Meal Headcount — ${formattedDate}**`);

    if (report.specialDay) {
        lines.push(specialDayLabel(report.specialDay.type));
    }

    lines.push("");

    if (report.meals.length === 0) {
        lines.push("No meals today (office closed or holiday).");
        return lines.join("\n");
    }

    for (const meal of report.meals) {
        const label = MEAL_LABELS[meal.type] || meal.type;
        lines.push(`${label}: ${meal.headcount}`);
    }

    lines.push("");
    lines.push(`In office: ${report.officeCount} · WFH: ${report.wfhCount}`);
    lines.push(`Total: ${report.totalUsers}`);
    lines.push("");

    if (report.specialDay?.note) {
        lines.push(`Note: ${report.specialDay.note}`);
    } else {
        lines.push("Regular working day.");
    }

    return lines.join("\n");
}

File: mhp-python/app/services/announcement_service.py
"""
Announcement service — generates a markdown-formatted daily announcement
suitable for copy-pasting into Discord or Slack.
Mirrors announcementService.js logic.
"""

from app.services.headcount_service import get_headcount_report
from app.services.special_day_service import get_special_day


def generate_announcement(date_str: str) -> str:
    report = get_headcount_report(date_str)
    special_day = get_special_day(date_str)

    lines = [f"## Meal Headcount — {date_str}"]

    if special_day:
        lines.append(f"\n> **{special_day['type']}**" + (f": {special_day['note']}" if special_day.get("note") else ""))

    if not report["meals"]:
        lines.append("\nNo meals today (office closed or holiday).")
        return "\n".join(lines)

    lines.append(f"\n**Office:** {report['officeCount']}  |  **WFH:** {report['wfhCount']}")
    lines.append("\n| Meal | Headcount |")
    lines.append("|------|-----------|")

    for meal in report["meals"]:
        lines.append(f"| {meal['type']} | {meal['headcount']} |")

    return "\n".join(lines)
